### PR TITLE
EZP-28881: Add a field to support "date object was trashed"

### PR DESCRIPTION
**Pull request by thiagocamposviana**
https://jira.ez.no/browse/EZP-28881

See also https://github.com/ezsystems/ezpublish-legacy/pull/1351

The eZ Publish trash can really fill and make it difficult to find something in the trash. It would be nice to be able to have a script that periodically clears out items that are older than "n days" (for example). In order to do this, we would need a field for objects in the trash that saves the "date object was trashed". It has been often requested by clients.

This pull request makes the changes from the ez publish legacy repository compatible with the ezpublish kernel.

For this pull request we should add the support for the date attribute the object was trashed, so users can implement their own command while there is not an official one.

### CODE REVIEW

**Comment by thiagocamposviana** [comment]
I am going to add the command line and update the necessary classes to make the new table column be updated just like in the legacy pull request.

**Comment by thiagocamposviana** [comment]
I have just updated the commit to include the script just like the commit for the legacy kernel.

**Comment by ernestob** [comment]
👍
This is a very useful feature.

**Comment by carlosmauri** [comment]
+1 great feature

**Comment by florianalexandre** [comment]
+1 on this one, we also have a customer requesting this feature.

**Comment by alongosz** [comment]
review ping @Nattfarinn as you've worked on Trash related thing recently.

**Comment by alongosz** [comment]
We also have upgrade scripts, currently in [/data/update](https://github.com/ezsystems/ezpublish-kernel/tree/master/data/update) for both MySQL and PostgreSQL. The format of the filename is `dbupdate-<latest_stable_version>-to-<next_version>.sql`. Right now (at the time of writing this) it would be `dbupdate-7.1.0-to-7.2.0.sql`

**Comment by thiagocamposviana** [comment]
@alongosz @Nattfarinn I have just updated the commit, could you review the changes?

**Comment by thiagocamposviana** [comment]
Note that the Mapper.php code  style has been fixed long time ago.

File: eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('ezplatform:trash-purge')
+            ->addArgument('trashed', InputArgument::OPTIONAL, 'If set, only objects that has been trashed for at least the specified amount of days will be purged.')
```
Comment (glye) [inline eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php]: This should be `trashed-days`, same as in the legacy PR.

File: eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            ->addOption(
+                'user',
+                'u',
+                InputOption::VALUE_OPTIONAL,
+                'eZ Platform username (with Role containing at least Content policies: read, versionread, edit, remove, versionremove)',
+                'admin'
+            )
+            ->setDescription('Empty eZ Publish trash. Permanently deletes all objects in the trash.');
```
Comment (glye) [inline eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php]: Not necessarily "all" anymore. Should be amended to explain that.

File: eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use eZ\Publish\API\Repository\TrashService;
+
+/**
+ * Console Command which empties eZ Publish trash.
+ */
+class TrashPurgeCommand extends ContainerAwareCommand
```
Comment (adamwojs) [inline eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php]: This command is registered in dependency injection container so it should extends `\Symfony\Component\Console\Command\Command` rather then`\Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand`

**Comment by thiagocamposviana** [comment]
I have updated the code and removed the unused variables, also used FQCN for the TrashService, updated the script description, updated the script trashed argument to trashed-days, used \Doctrine\DBAL\Connection for the handler variable.

The only thing I didn't change was extending \Symfony\Component\Console\Command\Command instead of \Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand because that will cause errors as long we do use the getContainer method.

**Comment by thiagocamposviana** [comment]
Any update on this one?

File: eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php (hunk 3 above)
Comment (blankse) [reply eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php]: Why don't you use `'@ezpublish.connection'` as the 2nd argument in the service definition and set the `$this->handler` in the constructor?

**Comment by thiagocamposviana** [comment]
I messed with the merge, going to fix it.

**Comment by thiagocamposviana** [comment]
Fixed the db conflicts.

**Comment by andrerom** [comment]
This is starting to look good, I'm missing improvements to expose option to sort on this with `TrashService::findTrashItems()`, but I'm well aware the filtering / sorting is not implemented in storage engine on that so this would be *far* beyond scope of this PR.

However should we simply call `TrashService::emptyTrash()` if no date is provided?  And if it is provides move the current logic to own methods which does that. with defined `trashedDays`.

**Comment by thiagocamposviana** [comment]
@andrerom at this point I think it is necessary to reduce the scope of this pull request as much as possible.

My problem with the request is that it is frustrating to get piece by piece of feedback spread out over months, I am willing to put in some more work but the current way this is being handled I would prefer to get this merged and then eZ could continue improving this overtime and that would make more sense as long you pointed out there are plans to improve then this could be reviewed and improved, but there are some customers asking about this feature for some months already.

File: eZ/Publish/API/Repository/Values/Content/TrashItem.php [hunk 4]
```diff
@@ -13,4 +13,10 @@
  */
 abstract class TrashItem extends Location
 {
+    /**
+     * Trashed timestamp.
+     *
+     * @var mixed Trashed timestamp.
+     */
+    protected $trashed;
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Values/Content/TrashItem.php]: I'm rather against timestamps in favor of `DateTime`. 

On the other hand, the one in `SPI\Persistence` should remain as a timestamp and be should be mapped to proper `DateTime`.

File: eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        $this->handler = $this->getContainer()->get('ezpublish.connection');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $trashedDays = (int) ($input->getArgument('trashed-days'));
+        $query = $this->handler->createSelectQuery();
```
[the remark this answers is not in the document]
Comment (andrerom) [reply eZ/Bundle/EzPublishCoreBundle/Command/TrashPurgeCommand.php]: 🗻

**Comment by SylvainGuittard** [comment]
That's a great feature. Thanks @thiagocamposviana for your contribution.
Regarding the priority of the queue, this is clear that we are in certification period for v2.2 right now and all our resources are focused on this.
After eZ Platform 2.2 release, it's totally possible.

**Comment by thiagocamposviana** [comment]
I am planning to split this pull request into two:

- One to add the new attribute to ezcontentobject_trash, making the changes to the affected code. This one should be ok according to the feedback so far.
- And another pull request for the command line related code changes.

It is clear the bottleneck of this pull request is the command BUT that part is not required to allow people to work on their own commands to purge items from the trash by date, so the most important part of this pull request for now for all customers is in fact the ezcontentobject_trash update.

At least I think I learned that it is very important to reduce the scope of a pull request, in this case I wanted to have the command but I didn't think about getting this split in two parts which would have worked way better for everybody, including people reviewing the code as long there would be less code per pull request to review and it would be easier to provide a full feedback in one step.

It could be even split in 3 pull requests, one for the basic attribute, one to provide the methods for the TrashService and finally the third for the command itself.

Maybe it could be helpful for ongoing pull request to ask contributors to split their pull requests in parts as a general rule of thumb.

**Comment by thiagocamposviana** [comment]
I have removed the command from this pull request, this can be implemented in a later pull request.

eZ/Publish/API/Repository/Values/Content/TrashItem.php is using now DateTime, while SPI\Persistence remains as a timestamp.

File: eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php [hunk 6]
```diff
@@ -45,6 +46,9 @@ public function createLocationFromRow(array $data, $prefix = '', Location $locat
         $location->depth = (int)$data[$prefix . 'depth'];
         $location->sortField = (int)$data[$prefix . 'sort_field'];
         $location->sortOrder = (int)$data[$prefix . 'sort_order'];
+        if (isset($data[$prefix . 'trashed'])) {
+            $location->trashed = new DateTime((int)$data[$prefix . 'trashed']);
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/Content/Location/Mapper.php]: From what I can see, this should be timestamp, the DateTime  conversion should be done in:
https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Repository/TrashService.php#L328-L345

Which is missing mapping of trashed currently, meaning api value has a null value atm.

**Comment by thiagocamposviana** [comment]
@andrerom I see the error from appveyor is not related, seems like there is an issue as you have pointed out in another pr:

https://github.com/ezsystems/ezpublish-kernel/pull/2357

> Seems the AppVeyor failure is unrelated (php not installing, might be temprary)

**Comment by thiagocamposviana** [comment]
I created an empty commit and rebased the branch just to trigger another AppVeyor build to see if the bug doesn't affect this PR.

**Comment by thiagocamposviana** [comment]
Looks like the workaround for the AppVeyor issue is to create an empty commit:

git commit --allow-empty -m "Empty test commit"

Then rebase:
git rebase -i HEAD~2

Instead of commenting out the empty commit, use squash.

This way it will create a new fake commit and when pushed it will trigger the AppVeyor build another time.

**Comment by thiagocamposviana** [comment]
Any update on this?

**Comment by thiagocamposviana** [comment]
Seems like due to the new conflicts I will need to review it again.

**Comment by thiagocamposviana** [comment]
I see actually there was a commit just some time ago that caused the conflicts here.

I am going to close this PR as long it has been said it would be reviewed last week, which it wasn't and now it is necessary again to keep solving conflicts indefinitely.

**Comment by glye** [comment]
@thiagocamposviana Sorry to hear that. This unfortunately happens sometimes also with issues that we work on internally. When a PR is big-ish and/or affects core classes or schema, it takes time to get reviews done, and you can be hit by other changes such as this schema change. Such is life on the bleeding edge :)

**Pull request closed without merging**